crucible/src/Lang/Crucible/Backend/Prove.hs

```diff
@@ -3,147 +3,452 @@ Module      : Lang.Crucible.Backend.Prove
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TupleSections #-}
 {-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE ScopedTypeVariables #-}
 
 module Lang.Crucible.Backend.Prove
-  ( ProofResult(..)
-  , proveGoal
-  , proveProofGoal
+  ( -- * Strategy
+    ProofResult(..)
+  , ProofConsumer(..)
+  , ProofStrategy(..)
+    -- ** Combiner
+  , SubgoalResult(..)
+  , Combiner(..)
+  , keepGoing
+  , failFast
+    -- ** Prover
+  , Prover(..)
+    -- *** Offline
+  , offlineProve
+  , offlineProveWithTimeout
+  , offlineProver
+    -- *** Online
+  , onlineProve
+  , onlineProver
+    -- * Proving goals
   , proveGoals
   , proveObligations
   , proveCurrentObligations
   ) where
 
 import           Control.Lens ((^.))
-import           Data.Functor.Const (Const(Const, getConst))
+import           Control.Monad.Catch (MonadMask)
+import           Control.Monad.Error.Class (MonadError, liftEither)
+import           Control.Monad.IO.Class (MonadIO(liftIO))
+import qualified Control.Monad.Reader as Reader
 
 import qualified What4.Interface as W4
 import qualified What4.Expr as WE
-import qualified What4.Solver.Adapter as WSA
+import qualified What4.Protocol.Online as WPO
+import qualified What4.Protocol.SMTWriter as W4SMT
 import qualified What4.SatResult as W4R
+import qualified What4.Solver.Adapter as WSA
 
 import qualified Lang.Crucible.Backend as CB
 import           Lang.Crucible.Backend.Assumptions (Assumptions)
-import           Lang.Crucible.Backend.ProofGoals (traverseGoalsWithAssumptions)
+import           Lang.Crucible.Utils.Timeout (Timeout, TimedOut)
+import qualified Lang.Crucible.Utils.Timeout as CTO
+
+-- | Local helper
+consumeGoals ::
+  -- | Consume an 'Assuming'
+  (asmp -> a -> a) ->
+  -- | Consume a 'Prove'
+  (goal -> a) ->
+  -- | Consume a 'ProveConj'
+  (a -> a -> a) ->
+  CB.Goals asmp goal ->
+  a
+consumeGoals onAssumption onGoal onConj = go
+  where
+  go (CB.Prove gl) = onGoal gl
+  go (CB.Assuming as gl) = onAssumption as (go gl)
+  go (CB.ProveConj g1 g2) = onConj (go g1) (go g2)
+
+-- | Local helper
+consumeGoalsWithAssumptions ::
+  forall asmp goal a.
+  Monoid asmp =>
+  -- | Consume a 'Prove'
+  (asmp -> goal -> a) ->
+  -- | Consume a 'ProveConj'
+  (a -> a -> a) ->
+  CB.Goals asmp goal ->
+  a
+consumeGoalsWithAssumptions onGoal onConj goals =
+  Reader.runReader (go goals) mempty
+  where
+  go :: CB.Goals asmp goal -> Reader.Reader asmp a
+  go =
+    consumeGoals
+      (\asmp gl -> Reader.local (<> asmp) gl)
+      (\gl -> Reader.asks (\asmps -> onGoal asmps gl))
+      (\g1 g2 -> onConj <$> g1 <*> g2)
+
+---------------------------------------------------------------------
+-- * Strategy
 
 -- | The result of attempting to prove a goal with an SMT solver.
 --
 -- The constructors of this type correspond to those of 'W4R.SatResult'.
+--
+-- * @sym@ is the symbolic backend, usually 'WE.ExprBuilder'
+-- * @r@ is the return type of the eventual 'ProofConsumer'
```

Comment:
`ProofResult` doesn't have an `r` type parameter. Should this be `t`?

Reply:
Whoops, yep!